src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -228,11 +230,91 @@ public ISqlConnectionFactory ConnectionFactory
-        /// null upon validation success</returns>
+        /// <param name="ownerUri">The URI of the connection</param>
+        /// <returns>true if a refresh is needed,
+        /// false if the uri cannot be found, refresh is not needed,
+        /// refresh token has already been requested, or if refresh fails </returns>
+
+        internal async Task<bool> TryRefreshAuthToken(string ownerUri)
+        {
+            ConnectionInfo connInfo;
+            if (this.TryFindConnection(ownerUri, out connInfo))
+            {
+                // If not an azure connection, no need to refresh token
+                if (connInfo.ConnectionDetails.AuthenticationType != "AzureMFA")
+                {
+                    return false;
+                }
+                else
+                {
+                    // Check if token is expired or about to expire
+                    if (connInfo.ConnectionDetails.ExpiresOn - DateTimeOffset.Now.ToUnixTimeSeconds() < maxTolerance)
+                    {
+                        // Check if the token is updating already, in which case there is no need to request a new one
+                        if (this.TokenUpdateUris.ContainsKey(ownerUri)) {
+                            return false;
+                        }
+                        var requestMessage = new RefreshTokenParams
+                        {
+                            AccountId = connInfo.ConnectionDetails.GetOptionValue("azureAccount", string.Empty),
+                            TenantId = connInfo.ConnectionDetails.GetOptionValue("azureTenantId", string.Empty),
+                            Provider = "Azure",
+                            Resource = "SQL",
+                            Uri = ownerUri
+                        };
+                        if (requestMessage.TenantId == null || requestMessage.TenantId == String.Empty)
+                        {
+                            Logger.Write(TraceEventType.Error, "No tenant in connection details when refreshing token for connection {ownerUri}");
+                            return false;
+                        }
+                        if (requestMessage.AccountId == null || requestMessage.AccountId == String.Empty)
+                        {
+                            Logger.Write(TraceEventType.Error, "No accountId in connection details when refreshing token for connection {ownerUri}");
+                            return false;
+                        }
+                        try
+                        {
+                            this.TokenUpdateUris.TryAdd(ownerUri, true);
+                            await this.ServiceHost.SendEvent(RefreshTokenNotification.Type, requestMessage);
+                        }
+                        catch (Exception ex)
+                        {
+                            Logger.Write(TraceEventType.Error, "Failed to send RefreshTokenNotification " + ex.Message);

```

Comment:
Just remove this try/catch. It would be pretty weird if it failed to send and would likely indicate other bigger problems - and regardless even if an error occurred since we need to refresh the token we wouldn't get intellisense anyways.